### PR TITLE
[query] move build system from Gradle to Mill

### PR DESCRIPTION
**Pull request by patrick-schultz**
## Mill setup

There is a mill wrapper script `millw` checked into the repo in the `hail/` subdirectory. You can either invoke it directly, as we did with `gradlew`, or copy it somewhere on your path and rename it `mill`. This way you can just run `mill ...`, and it will identify the correct version of mill to use for the project you're in and invoke that. For more details on the installation options, see the github for the wrapper script [here](https://github.com/lefou/millw).

To verify that it's working, and download the actual mill jar, run `./millw --version` (or `mill --version` if you put it on your path) from the `hail/` subdirectory. It should report version `0.11.6`.

## Mill from the command line

* `mill clean` - delete all output files (in `hail/hail/out`). Or only delete the output of one target, e.g. `mill clean test.compile`
* `mill compile` - compiles the root module (not including tests)
* `mill test.compile` - compiles tests (and, transitively, the rest of the root module)
* `mill test.test`, or for short `mill test` - run all tests. You can pass options to the test runner (TestNG currently), e.g.*  * `mill test -methods is.hail.expr.ir.CallFunctionsSuite.constructors` to run one test, or `mill test -threadcount 4 -parallel classes` to use 4 threads and parallelize over test classes
* `mill test.testOnly is.hail.expr.ir.CallFunctionsSuite` - run all tests in one or more specified classes. You can use `*` to match anything, e.g. `mill test.testOnly "*.CallFunctionsSuite"`, or `mill test.testOnly "is.hail.expr.ir.*"`. You can pass options to the test runner (TestNG currently) after a `--`, e.g. `mill test.testOnly "is.hail.expr.ir.*" -- -parallel classes`
* `mill __.testCached` - once the codebase is more modularized, will run tests on only modules whose dependencies have changed since the last test run
* `mill reformat` - runs scalafmt on all sources in the root module (currently that's all scala sources, but hopefully not for long). `mill __.reformat` runs scalaformat on all sources in all modules. `mill __.checkFormat` only checks for rule failures (we run this in ci)
* `mill fix` - runs scalafix on all sources in the root module (currently that's all scala sources, but hopefully not for long). `mill __.fix` runs scalafix on all sources in all modules. `mill __.fix --check` only checks for rule failures (we run this in ci). You can pass any options to scalafix, e.g. `mill fix --help`.
* `mill inspect` - see the docstring and dependencies for any target, e.g. `mill inspect compile`
* `mill ivyDepsTree` - show the tree of external dependencies of the root module, highlighting potential incompatibilities in transitive dependencies. `mill ivyDepsTree --withCompile --withRuntime` includes compile-only and runtime-only dependencies. Use `--whatDependsOn` to see an inverted tree showing how a transitive dependency is getting pulled in, e.g. `mill ivyDepsTree --withCompile --withRuntime --whatDependsOn org.slf4j:slf4j-api`
* `mill mill.scalalib.Dependency/showUpdates` - show outdated dependencies (we have a lot!)

See [here](https://mill-build.com/mill/Intro_to_Mill.html) and [here](https://mill-build.com/mill/Builtin_Commands.html) for more details.

## IntelliJ setup

To import from scratch:
* delete any `.idea` directories in the hail root, or `hail/` subdirectory (you could try skipping this, but you're on your own)
  * in the `hail/` subdirectory, you can also delete `.classpath`, `.gradle`, `.project`, `.settings/`, `build/` (we still use this for a few things, but most of it is dead gradle output, and it's safe to delete it all and start clean), `build.gradle`, `gradle/`, `gradlew`, `gradlew.bat`, `pgradle`, `settings.gradle`
* run `mill mill.bsp.BSP/install` to generate the `.bsp` config directory (bsp is the Build Server Protocol)
* In IntelliJ, go to File->Open, and choose the hail root directory
* When the project is open, go to File->Project Structure
  * in the Project pane, set an sdk (8 or 11), and set the language level to 8
  * in the Modules pane, delete the existing root module, click the plus sign -> Import Module, choose the `hail/` subdirectory, and choose "Import module from external model" and `BSP`
  * you should see a progress bar at the bottom as it imports the project
  * when it's done, quit and reopen IntelliJ. There should now be a bsp icon (two bars with two arrows between them) on the right, where the gradle elephant used to be. Just like before, sometimes you'll need to click the "reload" icon in there if things get wonky.
  * if it says "scalafmt configuration detected", go ahead and enable the formatter

## Metals setup

* delete any `.metals` directories
* open the hail repo in VSCode (even if you won't use VSCode, this seems to be the best way to get metals set up initially)
* it should ask you to import a Mill build
* when that finishes, at the bottom it should say it's connected to a Bloop build server. In general, I think using Mill as the BSP directly will work best, but I don't have much experience to say for sure. To switch, run `Metals: switch build server` from the command palette.

## Debug and release builds

As before, debug mode adds some (fairly expensive) checking to our native memory system. But now there are a few other differences:
* treat warnings as errors only in release mode, so you can still compile, run tests, etc. during development without fixing all warnings
* enable optimization in scalac only in release mode

The intention is that we use debug mode during development, and release mode ony for published artifacts, or performance profiling. Mill will use debug mode by default. To enable release mode, define `HAIL_RELEASE_MODE` in your environment. Note changing this will invalidate all mill intermediates and rebuild from scratch.

### CODE REVIEW

**Comment by patrick-schultz** [comment]
@chrisvittal: Ed has already used mill on his machine, and helped with the PR. It would be great to both get your eyes on the changes, as well as checking out the branch and testing the setup instructions above.

File: hail/Makefile [hunk 1]
```diff
@@ -21,23 +19,14 @@ HAIL_PYTHON3 ?= python3
...
 DOCTEST_TARGET ?= hail
 PYTEST_INTER_CLOUD_TARGET ?= test/hailtop/inter_cloud
 
-# not perfect, not robust to simdpp changes, but probably fine
-BUILD_DEBUG_PREFIX := build/classes/scala/debug
-JAR_SOURCES := $(shell git ls-files src/main)
-JAR_SOURCES += build.gradle
-JAR_TEST_SOURCES := $(shell git ls-files src/test)
-JAR_DEBUG_SOURCES := src/debug/scala/is/hail/annotations/Memory.java
-JAR_DEBUG_CLASSES := $(addprefix $(BUILD_DEBUG_PREFIX)/, $(JAR_DEBUG_SOURCES:src/debug/scala/%.java=%.class))
```
Comment (chrisvittal) [inline hail/Makefile:40]: We should keep the source/test dependencies on the jar targets, otherwise we may not rerun them when files change.

**Review by chrisvittal — CHANGES_REQUESTED** [review]
Honestly, this looks great. Did some testing of the commands and all seems well, though I don't use IDEA or VSCode, so I'm really not the best person to test them.

**Review by chrisvittal — DISMISSED** [review]

File: build.yaml [hunk 2]
```diff
@@ -844,9 +841,8 @@ steps:
     script: |
       set -ex
       cd /io/repo/hail
-      chmod 755 ./gradlew
-      time retry ./gradlew --version
-      export SPARK_VERSION="3.0.2" SCALA_VERSION="2.12.10"
+      time retry sh millw --version
+      export SPARK_VERSION="3.0.2" SCALA_VERSION="2.12.13" HAIL_RELEASE_MODE=1
```
Comment (patrick-schultz) [inline build.yaml:845]: @danking I only needed to change the scala version in azure, is this something we should manually test?

**Comment by patrick-schultz** [comment]
@chrisvittal I forgot about the native lib stuff. Going to dismiss your approval and work on that.

**Comment by chrisvittal** [comment]
Oops. I did too.

**Comment by patrick-schultz** [comment]
Wait, I don't know what I'm talking about. That's all at the make level.

**Comment by patrick-schultz** [comment]
Ah right, I deleted it

**Review by chrisvittal — APPROVED** [review]

**Comment by chrisvittal** [comment]
> So I just put back the dependencies on `native-lib-prebuilt`. Since that just calls make recursively, it would probably be better to let mill invoke make, but I didn't want to deal with that, and this is a pretty uncommon use case (I think).

That's sane to me, and avoids needing to deal with the make jobserver at the mill level. Feel free to un-WIP this whenever you're ready to merge.

File: build.yaml (hunk 2 above)
Comment (danking) [reply build.yaml:845]: I'm not concerned about a patch version change in Scala

**Pull request merged**